Lib/test/test_unparse.py

```diff
@@ -130,7 +130,32 @@ class Foo: pass
 
 class ASTTestCase(unittest.TestCase):
     def assertASTEqual(self, ast1, ast2):
-        self.assertEqual(ast.dump(ast1), ast.dump(ast2))
+        # Ensure the comparisons start at an AST node
+        self.assertIsInstance(ast1, ast.AST)
+        self.assertIsInstance(ast2, ast.AST)
+
+        # An AST comparison routine modeled after ast.dump(), but
+        # instead of string building, it traverses the two trees
+        # in lock-step.  Defining an inner function also prevents
+        # the `missing` value from polluting the "real" function
+        # signature or module namespace.
+        missing = object()
+        def traverse_compare(a, b):
+            if type(a) is not type(b):
+                self.fail("type(a) != type(b)")
+            if isinstance(a, ast.AST):
+                for field in a._fields:
+                    value1 = getattr(a, field, missing)
+                    value2 = getattr(b, field, missing)
+                    traverse_compare(value1, value2)
+            elif isinstance(a, list):
+                if len(a) != len(b):
+                    self.fail("len(a) != len(b)")
+                for node1, node2 in zip(a, b):
```

Comment:
You could also use the new `zip(a,b, strict=True)` and skip the previous test.

Reply:
I did originally, but using `self.fail()` from an except block results in a nested exception.  I don't know of a way to do `raise ... from None` without the `raise` :)